### PR TITLE
Add NISQA metric

### PR DESCRIPTION
**Pull request by philgzl**
## What does this PR do?

This PR adds a new metric [NISQA](https://github.com/gabrielmittag/NISQA) as suggested in #2464

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2792.org.readthedocs.build/en/2792/

### CODE REVIEW

**Review by SkafteNicki — COMMENTED** [review]
Hi @philgzl, thanks for this contribution to TM.
I am overall very happy with the PR already, it is nearly complete. If I could ask you to add some doc strings (just one liners is fine) to the different functions/methods in `src/torchmetrics/functional/audio/nisqa.py` that would be great for later maintaining the code.

File: src/torchmetrics/audio/nisqa.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    plot_upper_bound: float = 5.0
+
+    def __init__(self, fs: int, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        if not _LIBROSA_AVAILABLE or not _REQUESTS_AVAILABLE:
+            raise ModuleNotFoundError(
+                "NISQA metric requires that librosa and requests are installed. "
+                "Install as `pip install librosa requests`."
+            )
+        self.fs = fs
```
Comment (SkafteNicki) [inline src/torchmetrics/audio/nisqa.py:92]:
```suggestion
        if not isinstance(fs, int) or fs <= 0:
            raise ValueError(f"Argument `fs` expected to be a positive integer, but got {fs}")
        self.fs = fs
```

File: src/torchmetrics/functional/audio/nisqa.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+          networks", in Proc. ICASSP, 2019.
+        - [2] G. Mittag, B. Naderi, A. Chehadi and S. Möller, "NISQA: A deep CNN-self-attention model for
+          multidimensional speech quality prediction with crowdsourced datasets", in Proc. INTERSPEECH, 2021.
+
+    """
+    if not _LIBROSA_AVAILABLE or not _REQUESTS_AVAILABLE:
+        raise ModuleNotFoundError(
+            "NISQA metric requires that librosa and requests are installed. Install as `pip install librosa requests`."
+        )
+    model, args = _load_nisqa_model()
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/audio/nisqa.py:84]:
```suggestion
    model, args = _load_nisqa_model()
    if not isinstance(fs, int) or fs <= 0:
        raise ValueError(f"Argument `fs` expected to be a positive integer, but got {fs}")
```

File: src/torchmetrics/functional/audio/nisqa.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    myfile = requests.get(url)
+    with open(saveto, "wb") as f:
+        f.write(myfile.content)
+
+
+class _NISQADIM(nn.Module):
+    # main NISQA model definition
+    # ported from https://github.com/gabrielmittag/NISQA
+    # Copyright (c) 2021 Gabriel Mittag, Quality and Usability Lab
+    # MIT License
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/audio/nisqa.py:133]: @Borda we had copyright issue before, anything we need to do here?

File: tests/unittests/audio/test_nisqa.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+from torchmetrics.audio.nisqa import NonIntrusiveSpeechQualityAssessment
+from torchmetrics.functional.audio.nisqa import non_intrusive_speech_quality_assessment
+from torchmetrics.utilities.imports import _LIBROSA_AVAILABLE
+
+from unittests._helpers.testers import MetricTester
+
+if _LIBROSA_AVAILABLE:
+    import librosa
+else:
+    librosa = None
```
Comment (SkafteNicki) [inline tests/unittests/audio/test_nisqa.py:30]:
```suggestion

from unittests._helpers.testers import MetricTester
```
the import does not seem to be used during the testing?

Comment (philgzl) [reply tests/unittests/audio/test_nisqa.py:30]: True

File: src/torchmetrics/functional/audio/nisqa.py (hunk 3 above)
Comment (philgzl) [reply src/torchmetrics/functional/audio/nisqa.py:135]: The model weights are also under CC BY-NC-SA license

**Review by SkafteNicki — APPROVED** [review]

File: src/torchmetrics/functional/audio/nisqa.py (hunk 3 above)
Comment (Borda) [reply src/torchmetrics/functional/audio/nisqa.py:135]: that is a problem :(

Comment (Borda) [reply src/torchmetrics/functional/audio/nisqa.py:135]: we will need to get the authors to allow distribute it with apache licence

**Review by Borda — DISMISSED** [review]

File: src/torchmetrics/functional/audio/nisqa.py (hunk 3 above)
Comment (philgzl) [reply src/torchmetrics/functional/audio/nisqa.py:135]: Original code is not a pip installable package so can't really import as an optional dependency.

DNSMOS is similar: MIT license code and CC BY license weights. How was this managed? Or was this overlooked?

**Comment by Borda** [comment]
hote this is correct:

> If a package with an Apache License includes code that is under an MIT License, the combined project needs to respect both licenses. Here are the key implications:
> 
> * Compatibility: The Apache License 2.0 and MIT License are compatible, meaning code licensed under MIT can legally be included in a project under the Apache License.
>
> * License Notices: Since the MIT License requires preservation of its copyright and permission notice, these must be retained in the source files or documentation. In addition, the Apache License also requires its own licensing notice and attribution, so both notices should be clearly included.
> 
> * Patent Grant: The Apache License includes a patent grant, while the MIT License does not. If you're using MIT-licensed code in a project that you distribute under the Apache License, the Apache License's patent grant will only cover the parts of the code that are explicitly Apache-licensed, not the portions under MIT. However, this doesn’t prevent you from distributing the MIT code within the Apache-licensed package.
> 
> * Redistribution Terms: The resulting combined work can still be redistributed under the terms of the Apache License, as long as both licenses are respected in the documentation. Since both licenses are permissive, users are generally free to use, modify, and distribute the combined package, but the Apache License will apply to the package as a whole.
> 
> In practice, the combination works smoothly, and users of the package can use it as long as they comply with both license terms—chiefly by keeping both license texts intact and visible

.

**Comment by SkafteNicki** [comment]
@Borda added full copyright at the top of the affected file, so PR should be good to go now

**Review by lantiga — APPROVED** [review]
